Dispatchopts

RGunning opened the pull request:
Added option arguments to dispatch simple
set option in plite vsc

e.g.
dispatcher=dispatch-basic simple
dispatcher=dispatch-basic lsf -output this/dir

RGunning committed:
Update dispatch-basic

uncomment dispatc_lsf #l45 for LSF dispatcher

RGunning committed:
set default output directory

RGunning committed:
Update dispatch-basic

rewrite to App::Rad

RGunning committed:
Update dispatch-basic

RGunning committed:
Update dispatch-basic

RGunning committed:
Update dispatch-basic

RGunning committed:
Update Lite4.pm

allow options in dispatcher arguments

RGunning committed:
Update Lite4.pm

RGunning committed:
temp fix

temporarily commented out check dispatcher exists option #l277-278

RGunning committed:
Update Lite4.pm

RGunning committed:
forgot dollar

RGunning committed:
Dispatcher options

RGunning commented on lib/App/Pipeline/Lite4.pm:
```diff
@@ -268,21 +268,20 @@ sub run_pipeline {
-    my $pipeline_dir = $self->pipeline_dir->absolute->stringify;
-    ouch 'App_Pipeline_Lite4_Error', "Dispatcher does not exist at $dispatcher_exe" 
-     unless $dispatcher_exe->exists;
+  # ouch 'App_Pipeline_Lite4_Error', "Dispatcher does not exist at $dispatcher_exe" 
```
TODO: re-enable test if file exists

nw11 commented:
How much do you switch between dispatch-basic and lsf? I was thinking the best design for this, esp because more development might end up being done on the lsf dispatcher, is to have two dispatcher scripts dispatch-lsf and dispatch-basic. Then you just set them in the config file as you need (including setting the option to lsf for where it puts its log files). I don't think the external_dispatch function actually needs to be touched - it should just work fine. So basically, just paste the code for lsf stuff into dispatch-lsf.  I'm thinking this is the best strategy as I have some dispatcher scripts for job management systems besides lsf almost ready, and it seems more organised to do it this way, rather than to have a potentially massive file managing dispatch variants.

RGunning commented:
I never really switch.
It would make more sense to have them all the dispatchers in separate files.

However, for lsf, it would be a useful feature to pass the output directory through an option passed via plite vsc. This is simpler than requiring the user to edit the dispatcher script to change output directory.
It was because I wanted to do this that I started editing the external_dispatch function

RGunning commented on bin/dispatchers/dispatch-basic:
```diff
@@ -1,57 +1,157 @@
+	 	$args{-J} = $lsf_job_name;
+		$args{-R} = "span[hosts=1] select[mem>$mem] rusage[mem=$mem]" if defined($mem);
+		$args{-M} = $mem if defined($mem); # in KB, but we * by 1000 not 1024, since esub requires it to be less.
+		$args{-o} = $c->options->{'output'}.'/out-%J';
```
set output directory based on options.
Should probably set a default value

nw11 commented:
Yes it seems we are essentially talking about the same thing here.   

With something like this in the system config file (which is brought up using plite vsc): 

dispatcher=dispatch-lsf  -output this/dir

This would just be executed fine as per normal by external_dispatch. As it is just going to join the command ("dispatch-lsf -output this/dir") to the actual lsf commands file to get "dispatch-lsf -output this/dir some-lsf-cmd-graph-file.yaml", which is then run by system() in external_dispatch.

RGunning commented:
Not quite.

If you notice the line I commented out in the external_dispatch, the function is trying to test if the file at "dispatch-lsf -output this/dir” exists. This fails as only the first half of the command is a filepath, the second are options. My edits just separate out the file so I can test whether the file exists/executable

On 5 Aug 2015, at 17:26, nw11 <notifications@github.com> wrote:

> Yes it seems we are essentially talking about the same thing here.
> 
> With something like this in the system config file (which is brought up using plite vsc):
> 
> dispatcher=dispatch-lsf -output this/dir
> 
> This would just be executed fine as per normal by external_dispatch. As it is just going to join the command ("dispatch-lsf -output this/dir") to the actual lsf commands file to get "dispatch-lsf -output this/dir some-lsf-cmd-graph-file.yaml", which is then run by system() in external_dispatch.
> 
> —
> Reply to this email directly or view it on GitHub.
>

nw11 commented on lib/App/Pipeline/Lite4.pm:
```diff
@@ -268,21 +268,20 @@ sub run_pipeline {
-    my $dispatcher_exe=shift;
+    my $self = shift;
+    my $dispatcher_opt = shift;
+    my ($dispatcher_exe, $opts) = split / /, $dispatcher_opt;
```
probably do /\s+/ to make a bit more robust

nw11 commented:
ah yes, of course. Should have looked more closely! ok good, this should be fine  to merge, when the two dispatchers are separated

RGunning committed:
Create dispatch-lsf

RGunning committed:
Update dispatch-basic

RGunning commented:
I've split the dispatchers.

I should probably look at uncommenting lines l277-279 of external_dispatch which tests if the dispatcher file exists. Although l282 also does this when it tests if the dispatcher is executable

RGunning committed:
chmod

RGunning committed:
fix to dispatch cmd

->absolute->stringify was messing up the dispatch (it doesn't seem essential)

a dash was being added to the end of $dispatcher_opt

RGunning committed:
Update Lite4.pm

RGunning committed:
Update dispatch-lsf
